task: add ways to run a `LocalSet` from within a rt context

hawkw opened the pull request:
## Motivation

Currently, the only way to run a `tokio::task::LocalSet` is to call its
`block_on` method with a `&mut Runtime`, like

```rust
let mut rt = tokio::runtime::Runtime::new();
let local = tokio::task::LocalSet::new();
local.block_on(&mut rt, async {
  // whatever...
});
```

Unfortunately, this means that `LocalSet` doesn't work with the 
`#[tokio::main]`  and `#[tokio::test]` macros, since the `main` 
function is _already_ inside of a call to `block_on`.

## Solution

This branch adds a `LocalSet::run` method, which takes a future and
returns a new future that runs that future on the `LocalSet`. This
is analogous to `LocalSet::block_on`, except that it can be called in
an async context.

Additionally, this branch implements `Future` for `LocalSet`. Awaiting
a `LocalSet` will run all spawned local futures until they complete.
This allows code like

```rust
#[tokio::main] 
async fn main() {
    let local = tokio::task::LocalSet::new();

    local.spawn_local(async {
        // ...
    });

    local.spawn_local(async {
        // ...
        tokio::task::spawn_local(...);
        // ...
    });

    local.await;
}
```


I've also updated the `LocalSet` docs to show the usage with 
`#[tokio::main]` rather than with manually created runtimes, where
applicable.

Closes #1906 
Closes #1908 
Fixes #2057

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

carllerche commented on tokio/src/task/local.rs:
```diff
@@ -280,12 +318,70 @@ impl LocalSet {
+    ///
+    /// [`spawn_local`]: fn.spawn_local.html
+    /// [awaiting the local set]: #awaiting-a-localset
+    pub fn run<F>(&self, future: F) -> LocalFuture<F>
```
Maybe we should change this to `run_until` and could this be an `async fn`?

carllerche commented:
Should `LocalSet::block_on` be deprecated in favor of these fns?

carllerche commented:
You should also merge master and push (which will unblock CI).

Note that we have discovered that `cargo fmt` is broken. See how the rustfmt CI script has been updated.

hawkw commented:
@carllerche I believe I've addressed the review feedback!

One thought I had is that it _might_ be worth having an `async fn run_until_idle(&self)` on `LocalSet` that has the behaviour of `.await`ing a LocalSet but allows it to be reused afterwards. I'm not sure if this is worth having since it creates two separate APIs for the same behavior, and being able to `.await` a local set directly is nice. Any thoughts?

Matthias247 commented:
FWIW, `LocalSet` in general and 

> Additionally, this branch implements Future for LocalSet. Awaiting
a LocalSet will run all spawned local futures until they complete.

in particular look more and more on what we might want to do with #1879.

I think we can try to unify these at some point. It seems like the `LocalSet` use-cases could be implemented along `runtime.block_on(task::scope(spawn_stuff_here))`, where all tasks inside the scope are restricted to run a single thread (e.g. through options which are passed to a scope?)
